// accountdata/packet.js
var kb = require('keyedbits');
var validate = require('./validate.js');

function ClientPacket(call, remoteObj) {
    this.call = call;
    this.username = remoteObj['username'];
    this.password = remoteObj['hash'];
    this.device = remoteObj['device'];
    this.data = {};
    for (var key in remoteObj) {
        if (key != 'username' && key != 'hash' && key != 'device') {
            this.data[key] = remoteObj[key];
        }
    }
}

ClientPacket.prototype.validate = function() {
    if (!validate.validateCall(this.call, this.data)) {
        return false;
    }
    if (typeof this.username != 'string') {
        return false;
    }
    if (typeof this.device != 'string') {
        return false;
    }
    if (typeof this.password != 'object') {
        return false;
    }
    if (this.password.constructor.name != 'KBBuffer') {
        return false;
    }
    if (['ios'].indexOf(this.device) < 0) return false;
    
    return true;
}

ClientPacket.prototype.getAPIDomain = function() {
    return this.call.split('.')[0];
}

ClientPacket.prototype.getAPIName = function() {
    return this.call.split('.')[1];
}

function ServerPacket(object) {
    this.packet = object;
}

ServerPacket.prototype.encode = function (useBase64) {
    var buffer = new kb.buffer();
    var encoder = new kb.encode(buffer);
    encoder.encode(this.packet);
    if (useBase64) {
        return kb.base64.encode(buffer);
    } else {
        return buffer;
    }
}

exports.ClientPacket = ClientPacket;
exports.ServerPacket = ServerPacket;
