// webapi/handlers/account.js
var Handler = require('./base.js').Handler;
var db = require('accountdata').db;

function Account(aBuffer, contentType) {
    Handler.call(this, aBuffer, contentType);
}

Account.prototype = Object.create(Handler.prototype);

Account.prototype.isUnauthenticated = function(aCall) {
    return aCall == 'signup';
}

Account.prototype.setValues = function(callback) {
    callback(new Error('NYI'), null);
}

Account.prototype.getAccount = function(callback) {
    callback(null, this.account.toJSON());
}

Account.prototype.setHash = function(callback) {
    this.account.setField('hash', this.packet.data.new, function (err) {
        if (err) callback(err, null);
        else callback(null, {});
    });
}

Account.prototype.setEmail = function(callback) {
    this.account.setField('email', this.packet.data.email, function (err) {
        if (err) callback(err, null);
        else callback(null, {});
    });
}

Account.prototype.signin = function(callback) {
    callback(null, this.account.toJSON());
}

Account.prototype.signup = function(callback) {
    var username = this.packet.username;
    var hash = this.packet.password;
    var email = this.packet.data.email;
    var attributes = this.packet.data.attributes;
    db.account.signup(username, hash, email, attributes, function(err, acct) {
        if (err) {
            callback(err, null);
        } else {
            callback(null, acct.toJSON());
        }
    });
}

exports.handler = Account;
