// webapi/handlers/base.js
var db = require('accountdata').db;

/**
 * This is the abstract base class for an API object. The currently
 * defined API objects are puzzles, account, sessions, and image.
 */
function Handler(packet, contentType) {
    this.packet = packet;
    this.contentType = contentType;
    this.account = null;
}

Handler.prototype.isUnauthenticated = function(aCall) {
    return false;
}

Handler.prototype.handle = function(callback) {
    var call = this.packet.getAPIName();
    if (!this[call]) {
        return callback(new Error('Call not found on API handler'), null);
    }
    if (this.isUnauthenticated(call)) {
        this[call](callback);
    } else {
        var thisObj = this;
        db.account.findOne(this.packet.username, function (err, acct) {
            if (err) return callback(err, null);
            if (!acct.getField('hash').equals(thisObj.packet.password)) {
                return callback(new Error('Login incorrect'), null);
            }
            thisObj.account = acct;
            thisObj[call](callback);
        });
    }
}

exports.Handler = Handler;
